docker/docker.go

```diff
@@ -571,5 +584,71 @@ func Attach(ctx context.Context, config *container.Config, host *container.HostC
 		exit <- code
 	}()
 
+	monitorTtySize(c, res.ID)
+
 	return &resp, exit, nil
 }
+
+func getStdOutSize() (uint, uint) {
+	fd, isTerminal := term.GetFdInfo(os.Stdout)
+	if !isTerminal {
+		return 0, 0
+	}
+
+	ws, err := term.GetWinsize(fd)
+	if err != nil {
+		return 0, 0
+	}
+
+	return uint(ws.Height), uint(ws.Width)
+}
+
+func monitorTtySize(c *client.Client, containerID string) {
+	initTtySize(c, containerID)
```

Comment:
Is this initial call needed? For windows you're actually polling, and for non-windows your waiting for a signal. Maybe eventually needed only for non-windows?

Reply:
It's needed for all platforms. A user can resize terminal during the start of the container.

Reply:
What I mean is that in case of windows, if the first try in `initTtySize` fails, you then have to two goroutines doing essentially the same thing. Also if [this](https://github.com/src-d/engine/pull/364/files#diff-b715a530d535012e418a564a3eafedb9R610) is executed at a given time `t`, then it sleeps for `250ms`, in the meantime `[t, t+250]` the terminal could have been resized by the goroutine initialized [here](https://github.com/src-d/engine/pull/364/files#diff-b715a530d535012e418a564a3eafedb9R636), hence the previously calculated `prevH, prevW` could now be wrong.

Reply:
Those goroutines don't run at the same time. Init goroutine runs only for 50 ms and the second one waits for 250 ms. Multiple calls to `resizeTty` are safe so there is no problem with prev variables in terms of correctness of the code.

This implementation is heavily based on docker-cli implementation itself. It might have *super small* overhead in terms of trying to call resize functions when it's not really necessary. But on another hand, it's more "bullet-proof" and, in theory, should provide better UX in very weird cases.

Though I don't have a strong opinion on it. Please let me know if you still think it should be changed taking into account explanation above, I'll change.

Reply:
> Those goroutines don't run at the same time. Init goroutine runs only for 50 ms and the second one waits for 250 ms.

You're right, sorry I missed that.

> Multiple calls to resizeTty are safe so there is no problem with prev variables in terms of correctness of the code.

Sorry for the noise I've just realized that you're calling the resize whenever the sizes are different from the ones detected. I got confused with the initial state. 👍

Reply:
Sorry again, but I don't know why for some reason I have the impression that there's some sort of race-condition. I'm probably missing something else.

Let's say that at `t=0` the size is `(50, 50)`, then we have `(prevH, prevW) = (50, 50)`, then it sleeps for `250ms`. Let's say that during the initialization of the tty size I do a resize and, it fails 3 times, so in the end at `15ms` the current size is for example a `(100, 100)` and docker is aware of this new size hence it gives a better UX. Before waking up from the sleep of `250ms` I resize again exactly at `(50, 50)` let's say at `100ms` (the polling hasn't started yet). At `250ms` starts the polling but [this condition](https://github.com/src-d/engine/pull/364/files#diff-b715a530d535012e418a564a3eafedb9R615) is actually never met, so the UX is not optimized for the `(100, 100)` size.